### PR TITLE
TPM simulator -add Optional macro to derive EPS seed from device hardware parameters(linux only)

### PR DESCRIPTION
**Pull request by mohandatla**
Read MAC address (6 bytes) + Disk serial number generally 10 bytes and use that as entropy when USE_DEVICE_PERSISTENT_IDENTITY is set in configure.ac
todo: By default added libudev dependency, need to modify the configure.ac to give an option to the user while building it otherwise it should just build the default DEBUG behavior.

### CODE REVIEW

**Comment by amarochk** [comment]
Hi Mohan,

After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:

```C
    TPM2B_STRING(EPS_CREATION, "EPS Creation");
    DRBG_STATE           rand;
    TPM2B                deviceID;
    TPM_RC result = GetDeviceID(deviceID);
    result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
    DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
```

PRIMARY_SEED_CREATION definition should go to Global.h.

I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
image

-Andrey

File: Makefile.am [hunk 1]
```diff
@@ -55,7 +55,7 @@ TPMCmd_Simulator_src_tpm2_simulator_CFLAGS = $(EXTRA_CFLAGS) $(PLATFORM_INC) \
 # the weird / duplicate static library is necessary for dealing with the
 # circular dependency beetween libplatform and libtpm
 TPMCmd_Simulator_src_tpm2_simulator_LDADD = $(libplatform) $(libtpm) \
-    $(libplatform) $(LIBCRYPTO_LIBS) $(PTHREAD_LIBS) $(LIBUDEV_LIBS)
+    $(libplatform) $(LIBCRYPTO_LIBS) $(PTHREAD_LIBS) @ADDIOTIONAL_LIBS@
```
Comment (paulmon) [inline Makefile.am]: nitpick: ADDITIONAL (spelling)

**Comment by mohandatla** [comment]
> deliberation

Thank you for you suggestion and offline discussion. As per your suggestion, I would make changes for EPS seed to generate using device_id.

File: Makefile.am (hunk 1 above)
Comment (mohandatla) [reply Makefile.am]: > nitpick: ADDITIONAL (spelling)

Thanks Paul, will fix it in my next iteration.

**Comment by mohandatla** [comment]
> Hi Mohan,
> 
> After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:
> 
> ```c
>     TPM2B_STRING(EPS_CREATION, "EPS Creation");
>     DRBG_STATE           rand;
>     TPM2B                deviceID;
>     TPM_RC result = GetDeviceID(deviceID);
>     result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
>     DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
> ```
> 
> PRIMARY_SEED_CREATION definition should go to Global.h.
> 
> I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
> 
> -Andrey

Thank you for you suggestion and offline discussion. As per your suggestion, I would make changes for EPS seed to generate using device_id. Current PR.
I would also do separate PR for PPS & SPS to be in sync to make all three seeds are in sync & persistent.  Second PR. We can discuss later about taking this up or not.

File: configure.ac [hunk 2]
```diff
@@ -43,7 +43,19 @@ AM_INIT_AUTOMAKE([foreign subdir-objects])
...
+dnl By enabling this feature tpm simulator gets device persistent entropy derived from hardware.
+dnl It is enabled only for linux devices.
+dnl Note that the entropy is not derived from secure hardware source.
+
+AC_ARG_ENABLE(usedevicepersistententropy,
+    AS_HELP_STRING([--enable-usedevicepersistententropy],
+    [tpm simulator gets device persistent entropy derived from hardware. Entropy is not derived from secure source.]))
+
 PKG_CHECK_MODULES([LIBCRYPTO], [libcrypto])
+AS_IF([test "x$enable_usedevicepersistententropy" = "xyes"], [
```
Comment (jslobodzian) [inline configure.ac]: >x [](start = 13, length = 1)

why x?

Comment (jslobodzian) [reply configure.ac]: Just seems a little strange...  comparing something to "yes" which makes sense to the code reader vs.  xyes which looks like "xeyes" which is an X Windows management system utilitiy that follows your mouse cursor around.

Also...there is the bit about appending an x to the left side of $enable....  which is hard to see and looks funny

---
In reply to: [295547827](https://github.com/ms-iot/ms-tpm-20-ref/pull/1#discussion_r295547827) [](ancestors = 295547827)

**Comment by mohandatla** [comment]
> > Hi Mohan,
> > After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:
> > ```c
> >     TPM2B_STRING(EPS_CREATION, "EPS Creation");
> >     DRBG_STATE           rand;
> >     TPM2B                deviceID;
> >     TPM_RC result = GetDeviceID(deviceID);
> >     result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
> >     DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
> > ```
> > 
> > 
> > PRIMARY_SEED_CREATION definition should go to Global.h.
> > I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
> > -Andrey
> 
> Thank you for you suggestion and offline discussion. As per your suggestion, I would make changes for EPS seed to generate using device_id. Current PR.
> I would also do separate PR for PPS & SPS to be in sync to make all three seeds are in sync & persistent. Second PR. We can discuss later about taking this up or not.



> Hi Mohan,
> 
> After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:
> 
> ```c
>     TPM2B_STRING(EPS_CREATION, "EPS Creation");
>     DRBG_STATE           rand;
>     TPM2B                deviceID;
>     TPM_RC result = GetDeviceID(deviceID);
>     result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
>     DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
> ```
> 
> PRIMARY_SEED_CREATION definition should go to Global.h.
> 
> I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
> 
> -Andrey

Andrey, updated PR to derive the EPS seed from device id. Added code to derive the PPS & SPS from device id as well. I can make the PPS & SPS seed derivation as a separate optional feature in case if you think is totally against the TCG standards or other reason.

File: TPMCmd/tpm/src/subsystem/Hierarchy.c [hunk 3]
```diff
@@ -64,6 +64,7 @@ HierarchyPreInstall_Init(
 #else
     CryptRandomGenerate(gp.EPSeed.t.size, gp.EPSeed.t.buffer);
 #endif
+
```
Comment (mohandatla) [inline TPMCmd/tpm/src/subsystem/Hierarchy.c]: somehow missed, will do in my final interation.

**Pull request merged**